adal/src/main/java/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -141,28 +145,77 @@ public AuthenticationContext(Context appContext, String authority, boolean valid
      * @param tokenCacheStore Cache {@link ITokenCacheStore} used to store
      *                        tokens. Set to null if you don't want cache.
      */
-    public AuthenticationContext(Context appContext, String authority,
-                                 ITokenCacheStore tokenCacheStore) {
+    public AuthenticationContext(@NonNull final Context appContext,
+                                 @NonNull final String authority,
+                                 @Nullable final ITokenCacheStore tokenCacheStore) {
         initialize(appContext, authority, tokenCacheStore, true, false);
     }
 
-    private void initialize(Context appContext, String authority, ITokenCacheStore tokenCacheStore,
-                            boolean validateAuthority, boolean defaultCache) {
+    private void initialize(@NonNull final Context appContext,
+                            @NonNull final String authority,
+                            @Nullable final ITokenCacheStore tokenCacheStore,
+                            final boolean validateAuthority,
+                            final boolean defaultCache) {
         if (appContext == null) {
             throw new IllegalArgumentException("appContext");
         }
+
         if (authority == null) {
             throw new IllegalArgumentException("authority");
         }
+
         mBrokerProxy = new BrokerProxy(appContext);
+
         if (!defaultCache && !mBrokerProxy.canUseLocalCache(authority)) {
             throw new UnsupportedOperationException("Local cache is not supported for broker usage");
         }
+
         mContext = appContext;
         checkInternetPermission();
         mAuthority = extractAuthority(authority);
         mValidateAuthority = validateAuthority;
-        mTokenCacheStore = tokenCacheStore;
+
+        if (null != tokenCacheStore) {
+            mTokenCacheStore = wrapCache(tokenCacheStore);
+        }
+    }
+
+    private ITokenCacheStore wrapCache(@NonNull final ITokenCacheStore originalCache) {
+        return new ITokenCacheStore() {
+            @Override
+            public TokenCacheItem getItem(final String key) {
```

Comment:
@AdamBJohnsonx This is the suggestion? For each of these overridden methods? (Just confirming I understood your feedback)

```suggestion
            public synchronized TokenCacheItem getItem(final String key) {
```

Reply:
Yeah.  The main goal would be that we want removeAll and setItem to have a defined behavior.

Reply:
ACK. Will push

Reply:
Applied in 7a26dbee681f6ab9fd5fe8a7d10ff13d586414fe